src/System.IO.Ports/src/System/IO/Ports/SerialStream.Unix.cs

```diff
@@ -381,115 +353,131 @@ public override int Read(byte[] array, int offset, int count)
-            if (count == 0) return 0; // return immediately if no bytes requested; no need for overhead.
-
-            if (timeout != 0)
+            using (CancellationTokenSource cts = GetCancellationTokenSourceFromTimeout(timeout))
             {
-                Interop.Sys.PollEvents events = Interop.Sys.PollEvents.POLLNONE;
-                Interop.Sys.Poll(_handle, Interop.Sys.PollEvents.POLLIN | Interop.Sys.PollEvents.POLLERR, timeout,out events);
+                Task<int> t = ReadAsync(array, offset, count, cts?.Token ?? CancellationToken.None);
 
-                if ((events & (Interop.Sys.PollEvents.POLLERR | Interop.Sys.PollEvents.POLLNVAL)) != 0)
+                try
                 {
-                    throw new IOException();
+                    return t.GetAwaiter().GetResult();
                 }
-
-                if ( (events & Interop.Sys.PollEvents.POLLIN) == 0)
+                catch (OperationCanceledException)
                 {
                     throw new TimeoutException();
                 }
             }
+        }
 
-            int numBytes;
-            fixed (byte* bufPtr = array)
-            {
-                numBytes = Interop.Sys.Read(_handle, bufPtr + offset, count);
-            }
+        public override int EndRead(IAsyncResult asyncResult)
+            => EndReadWrite(asyncResult);
 
-            if (numBytes < 0)
-            {
-                if (Interop.Error.EWOULDBLOCK == Interop.Sys.GetLastError())
-                {
-                    throw new TimeoutException();
-                }
-                throw new IOException();
-            }
+        public override Task<int> ReadAsync(byte[] array, int offset, int count, CancellationToken cancellationToken)
+        {
+            CheckReadWriteArguments(array, offset, count);
 
-            return numBytes;
+            if (count == 0)
+                return Task<int>.FromResult(0); // return immediately if no bytes requested; no need for overhead.
+
+            Memory<byte> buffer = new Memory<byte>(array, offset, count);
+            SerialStreamIORequest result = new SerialStreamIORequest(cancellationToken, buffer);
+            _readQueue.Enqueue(result);
+
+            return result.Task;
         }
 
-        public override void Write(byte[] array, int offset, int count)
+        public override Task WriteAsync(byte[] array, int offset, int count, CancellationToken cancellationToken)
         {
-            Write(array, offset, count, WriteTimeout);
+            CheckWriteArguments(array, offset, count);
+
+            if (count == 0)
+                return Task.CompletedTask; // return immediately if no bytes to write; no need for overhead.
+
+            Memory<byte> buffer = new Memory<byte>(array, offset, count);
+            SerialStreamIORequest result = new SerialStreamIORequest(cancellationToken, buffer);
+            _writeQueue.Enqueue(result);
+            return result.Task;
         }
```

Comment:
We should override the Read/WriteAsync methods that take a `{ReadOnly}Memory<byte>` as well.

Reply:
@stephentoub this currently targets netstandard so those are not present there yet